lib/membrane/silence_generator.ex

```diff
@@ -0,0 +1,76 @@
+defmodule Membrane.SilenceGenerator do
+  @moduledoc """
+  Element responsible for generating silence as raw audio.
+  """
+
+  use Membrane.Source
+
+  alias Membrane.{Buffer, Time}
+  alias Membrane.Caps.Audio.Raw
+
+  def_options caps: [
+                type: :struct,
+                spec: Raw.t(),
+                description: "Audio caps"
+              ],
+              duration: [
+                type: :integer,
+                spec: Time.t(),
+                description: "Duration of the silence"
+              ],
+              frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output, caps: Raw
+
+  @impl true
+  def handle_init(opts) do
+    state =
+      opts
+      |> Map.from_struct()
+      |> Map.put(:passed_time, 0)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, %{caps: caps} = state) do
+    {{:ok, caps: {:output, caps}}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _ctx, %{caps: caps} = state) do
+    time = Raw.bytes_to_time(size, caps)
+    do_handle_demand(time, state)
+  end
+
+  def handle_demand(:output, buffers, :buffers, _ctx, state) do
+    %{caps: caps, frames_per_buffer: frames_per_buffer} = state
+
+    time = buffers * Raw.frames_to_time(frames_per_buffer, caps)
+    do_handle_demand(time, state)
+  end
+
+  defp do_handle_demand(time, state) do
+    %{caps: caps, duration: duration, passed_time: passed_time} = state
+
+    if passed_time + time < duration do
+      buffer = %Buffer{payload: Raw.sound_of_silence(caps, time)}
+      state = %{state | passed_time: passed_time + time}
```

Comment:
If time is no way to provide a buffer that is exactly that long, the rounding will cause the passing time calculation to be invalid. I think you should recalculate the time that was actually pushed.
The error can be seen e.g. when you take 537 ms of audio with 44100Hz sampling
```
iex(33)> Raw.sound_of_silence(caps, 537 |> Membrane.Time.milliseconds) |> byte_size |> Raw.bytes_to_time(caps) 
537006802
iex(34)> 537 |> Membrane.Time.milliseconds
537000000
```

You can make a test out of it

Reply:
You're right, I've missed that the time has already been rounded before `do_handle_demand`.